storage-proofs/src/rational_post.rs

```diff
@@ -0,0 +1,473 @@
+        Ok(PublicParams {
+            sector_size: sp.sector_size,
+            challenges_count: sp.challenges_count,
+        })
+    }
+
+    fn prove<'b>(
+        _pub_params: &'b Self::PublicParams,
+        pub_inputs: &'b Self::PublicInputs,
+        priv_inputs: &'b Self::PrivateInputs,
+    ) -> Result<Self::Proof> {
+        assert_eq!(
+            pub_inputs.challenges.len(),
+            pub_inputs.commitments.len(),
+            "mismatched challenges and commitments"
+        );
+        let challenges = pub_inputs.challenges;
+
+        let proofs = challenges
+            .iter()
+            .zip(pub_inputs.commitments.iter())
+            .map(|(challenge, commitment)| {
+                let challenged_sector = challenge.sector;
+                let challenged_leaf = challenge.leaf;
+
+                let tree = priv_inputs.trees[challenged_sector as usize];
+                if commitment != &tree.root() {
+                    return Err(Error::InvalidCommitment);
+                }
+
+                Ok(MerkleProof::new_from_proof(
+                    &tree.gen_proof(challenged_leaf as usize),
+                ))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        Ok(Proof {
+            inclusion_proofs: proofs,
+        })
+    }
+
+    fn verify(
+        pub_params: &Self::PublicParams,
+        pub_inputs: &Self::PublicInputs,
+        proof: &Self::Proof,
+    ) -> Result<bool> {
+        let challenges = pub_inputs.challenges;
+
+        if challenges.len() != pub_inputs.commitments.len() as usize {
+            return Err(Error::MalformedInput);
+        }
+
+        if challenges.len() != proof.inclusion_proofs.len() {
+            return Err(Error::MalformedInput);
+        }
+
+        // validate each proof
+        for ((merkle_proof, challenge), commitment) in proof
+            .inclusion_proofs
+            .iter()
+            .zip(challenges.iter())
+            .zip(pub_inputs.commitments.iter())
+        {
+            let challenged_leaf = challenge.leaf;
+
+            // validate the commitment
+            if merkle_proof.root() != commitment {
+                return Ok(false);
+            }
+
+            // validate the path length
+            if graph_height(pub_params.sector_size as usize / NODE_SIZE)
+                != merkle_proof.path().len()
+            {
+                return Ok(false);
+            }
+
+            if !merkle_proof.validate(challenged_leaf as usize) {
+                return Ok(false);
+            }
+        }
+
+        Ok(true)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Challenge {
+    pub sector: u64,
+    pub leaf: u64,
+}
+
+/// Rational PoSt specific challenge derivation.
+pub fn derive_challenges(
+    challenge_count: usize,
+    sector_size: u64,
+    sector_count: u64,
+    seed: &[u8],
+    faults: &[u64],
+) -> Vec<Challenge> {
+    // TODO: ensure sorting of faults
```

Comment:
SUGGESTION: Since your code requires sorted stuff, I _highly_ suggest that you introduce a new type to replace `&[u64]` which represents a set of sorted unsigned, 64-bit integers. Something like:

```
pub struct SortedSectorIds<'a> {
    ids: &'a [u64] // not pub
}
```

also, create

```
impl From<&[u64]> for SortedSectorIds {
    fn from(ns: &[u64]) -> Self {
        // sort ns
    }
}

impl From<SortedSectorIds> for &[u64] {
    fn from(x: SortedSectorIds) -> Self {
        x.ids
    }
}
```

Reply:
I think I would like to introduce a general `pub type SectorId = u64` into the code base, and as part of that introduce this kind of lists. But I would like to defer this to another PR.

Reply:
> pub type SectorId = u64

Type aliases don't provide value beyond documentation. If you are going to introduce a `SectorId` type (or change the one we have), please [create a newtype](https://doc.rust-lang.org/1.0.0/style/features/types/newtype.html), as per the docs. 

Motivation: In the above case, any `u64` can be used where a `SectorId` is required (as they're the same type).

Reply:
> But I would like to defer this to another PR.

If you're going to do that, would you please create an issue in rust-fil-proofs and link to that issue in a TODO-comment?

Reply:
https://github.com/filecoin-project/rust-fil-proofs/issues/839

Reply:
Thank you sir